### ISSUE TITLE
[question] How to manage requirements on dependency versions without consuming them

### ISSUE THREAD

**Issue by aander80** [conan-io/conan#17399]
### What is your question?

Hi!

I have a question regarding dependency management.

Let's say that we have two libraries: `liba` and `libb` (static libraries). They are linked together in `app`. The two libraries are communicating with each other, but are not directly calling code in the other library; this is done by automatically generating a communication layer in `app`. However, each library is dependent on the other providing signals that the communication layer can provide. Both libraries read from and write to each other. Hence, `liba` has dependencies to `libb` and vice versa, although they are not required to build (compile) each individual `lib*` conan package. We don't configure these dependencies as `requires` in the conanfiles in `liba` and `libb`, since that would result in circular dependencies and error messages in Conan.

The problem that arises is if we use version ranges for `liba` and `libb` in `app`. For simplicitly, let's assume `app` defines

```requires = "liba/[>=0, include_prerelease]", "libb/[>=0, include_prerelease]"```

Newer versions of both libraries will be picked up automatically, but they might each require a specific version of the other to link to get the final binary. Let's assume we commit a change to `liba` with version `1.2.3` that requires at least `libb/2.3.4` for `app` to build. We could of course commit a new change to the `app` repo which sets

```requires = "liba/[>=1.2.3, include_prerelease]", "libb/[>=2.3.4, include_prerelease]"```

However, this would only solve the issue in the future, but for older commits of `app` the dependencies between `liba` and `libb` would not be highlighted. Additionally, if we need to update `app` for every change to `liba` which depends on a specific version of `libb` then we have cyclic dependencies between the commits in `liba` and `app` which we would like to avoid.

My question is: can we use conan to specify the dependencies between `liba` and `libb`, but the dependency specifications are not applied until the graph for `app` is built? I am thinking of something along the lines of `generator_info`, where you specify something in `package_info()` in `liba` and `libb`, but it is only intended for the consumers of the packages.

I am thinking of two potential behaviors:
* An additional field in `package_info()` which *always* adds requirements to the consumer, something along the lines of
  ```self.cpp_info.transitive_requires = "libb/[>=2.3.4, include_prerelease]"```
  With this, you could theoretically use `requires = "liba/[>=0, include_prerelease=True]"` in `app` and `libb` would still be injected as a requirement for `app`.
* An additional field in `package_info` which *conditionally* sets the version specification for the rest of the build graph, but only if that package name is added in the dependency graph by another package. In our case, we would configure something along the lines of
  ```self.cpp_info.conditional_requires = "libb/[>=2.3.4, include_prerelease]"```

What are you thoughts on my problem and suggestions?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @aander80 

It seems you have a complicated problem at hand.

> An additional field in package_info() which always adds requirements to the consumer, something along the lines of
self.cpp_info.transitive_requires = "libb/[>=2.3.4, include_prerelease]"
With this, you could theoretically use requires = "liba/[>=0, include_prerelease=True]" in app and libb would still be injected as a requirement for app.

the ``package_info()`` is information to consumers of the package that is only evaluated once the full dependency graph has been resolved, computing all the versions, after the ``package_id`` of every package has been computed and after the binaries of the packages have been downloaded to disk (because some ``package_info()`` need the actual binaries on disk).
As such, it is impossible to have anything in ``package_info`` that affects how versions are resolved.

I am probably not understanding the problem completely.
I would say that this is solved by standard versioning. Having:
```
requires = "liba/[>=0, include_prerelease]", "libb/[>=0, include_prerelease]"
```
Is mostly not correct in most cases. Specially when there are future changes that makes libraries incompatible. So if you start with

```
requires = "liba/[>=1 <2, include_prerelease]", "libb/[>=1 <2, include_prerelease]"
```
Then, if something changes in liba and libb that introduces incompatibilities, you can control that by bumping their versions accordingly above ``2``, so all previous versions of ``app`` remain working without problems.

Basically version-ranges cannot "lie", if a recipe says that some versions are ok, they are ok, it is not valid in the future to bring a new version that would break that consumer and say "I don't want to be used by that consumer anymore" without bumping its version above the limit of the range of that consumer.

**Comment by aander80** [comment]
Thanks for your reply! When thinking about it I think we need to think more about how to solve the traceability and dependency problem when we have version ranges. There is of course an issue when changes to the libraries are not themselves breaking, but only when built into `app`, and then only depending on the order in which changes are merged and packages are deployed. Perhaps this will not be an issue for us long-term; if we do want to check out an old commit in `app` and make new release based on that, we can use lockfiles from the old release and base our dependencies on that.

There is, however, one thing that complicates things.

One thing we are considering is to commit lockfiles to each repo (we currently do not need to build for multiple profiles/configurations so we only have one set of dependencies per library/app). This will preserve a stable set of dependencies for all repos and simplify releases; you can make releases from any commit without having to explicitly lock down dependencies, while also allowing continuous integration through version ranges in `app`.

One issue with this approach is that we need to build each new commit in each repo twice: once with the locked dependencies from the lockfile, and once from the non-locked dependencies using only the version ranges. Assume that we have `liba/1.2.3` and `libb/2.3.4` in the lockfile in `app`, but `app` only locks the major version of each dependency (`liba/[~1]` and `libb/[~2]`). The latest version of each library is `liba/1.4.0` and `libb/2.7.3`. `liba/1.2.3` + `libb/2.3.4` work together, as does `liba/1.4.0` + `libb/2.7.3`.

However, if we now bump the version of `liba` in the lockfile so that it now contains `liba/1.4.0` + `libb/2.3.4`, then we might run into issues if they are not compatible. Assume `liba/1.4.0` is dependent on changes that came in in `libb/2.6.0`. Then, we would have to step back and see which version of `libb` we would also need to bump in the `app` lockfile for the build to pass. This could be tedious when the number of libraries goes to 50 or 100, especially when you expect the dependencies to build properly since they passed CI tests before.

Moreover, if we want to add additional CI checks to make sure that if we only update the version of one dependency in the `app` lockfile the build would still pass (i.e. only update the version in the lockfile from `liba/1.2.3` to `libb/1.4.0`), then we might lack the metadata in `liba` that it now requires another version of `libb` to build properly. I suppose this does not really find all the compatibility issues if we have updated to `libb/2.7.3` in the `app` lockfile already.

As I am writing this, I am not sure how much responsibility Conan as a tool should have in our case. We can of course add more checks and make sure that people are bumping major versions whenever they introduce breaking changes (and we need to figure out which types of these breaking changes there are so that we can catch them properly), but this also means that if we are locking major versions of dependencies in `app`, then we need to go in to `app` and make changes much more often to make sure the new major versions of `liba` and `libb` are continuously integrated into `app`.

What are your thoughts on this? Do you have a suggestion for deciding which set of dependencies we build with in CI builds to ensure stability? I am personally starting to lean towards keeping it the way we are moving towards (two builds: one with locked dependencies through lockfiles, one with unlocked dependencies using only version ranges) and generating warnings or errors if there are changes that we consider breaking in `app`. The more I think about this I feel like there is no overall good solution but rather different solutions that are more or less bad for our use-cases (at which point we of course strive for the least bad option :) ).

**Comment by memsharded** [comment]
I am not sure if I have already shared this with you: https://github.com/conan-io/docs/pull/3799

Also we have prepared a blog post: https://github.com/conan-io/conan-io.github.io/pull/277

It will be live next week, but you can read it in source form (or generate the docs with sphinx ``make html`` from my branch). It will take some time to read, it is a full tutorial, with hands-on exercises, figures and lots of explanations and discussions. The code to run it is also in another open PR to the ``examples2`` repo.

I think there are some important insights there, like the package promotions and the "products pipeline", which seems very connected to what you are describing, and the final goal of having both stability for developers and CI and the ability to move forward with new versions. 

Please have a look and let me know your thougths.

**Comment by aander80** [comment]
Sorry for the late reply! I have had a look at it, and there are some interesting aspects for us to consider, particularly the difference between the `packages`, `products` and `develop` and how packages are promoted only on specific builds (game + mapviewer), compared to treating each package individually, perhaps this can be of some use to us to solve the issues I described above. My schedule is a bit hectic before Christmas so I am not sure I will have time to dig deeper into it, but it was a very interesting read and I think it is definitely useful to showcase the difficulties that can occur with layered integration like this!

**Comment by memsharded** [comment]
For reference, the CI tutorial is now live in https://docs.conan.io/2/ci_tutorial/tutorial.html

Maybe we can close the ticket by now if it doesn't need an immediate further response, and maybe re-start the conversation (re-open or create a new ticket) if you want after Christmas based on the info from the tutorial?

**Comment by aander80** [comment]
Sure we can do that, thanks!
